### PR TITLE
[Feat] Dynamic and trailing stoploss

### LINKED ISSUE #20: Write first Unit Tests
After completing #19 , Unit tests need to be added. In consultation with Tijs, you are going to write automated tests for our engine! In the end, we strive for a code coverage of 75%.

### LINKED ISSUE #147: Make custom stoploss method work with current implementation
After the adjustments for overall backtesting speed, the implementation on Custom StopLoss function in the strategy.py file has not been implemented correctly. This currently overrides stoploss for ALL trades, while we actually want it to set a custom stoploss for EVERY trade. 

Please get in touch with Tijs if things are unclear. 
 
Things to do:
- Revise tradingmodule.py line 87/88 (open_trade_tick)
- Revise backtesting.py line 88/89 (populate_indicators)
- Make sure the Stoploss is put into data_dict in backtesting.py (start_backtesting)
- Revise tradingmodule.py method check_stoploss_open_trade to start using either custom stoploss if set, else use configured SL from config.

### PR DESCRIPTION
**Pull request by dema-trading-ai**
# Fixes

Fixes #147 
Fixed #20 

# What has been changed?

* Fixed wrong calculation of price at trade close (see explanation above)
* Restructured stoploss calculations
* Added trailing stoploss and dynamic stoploss
* Added 'stoploss-type' to config.json so users can switch between different types
* Added my_strategy_advanced.py to show how dynamic stoploss should be used
* Added first unit test

# Examples

* Users now have the choice to choose three different options as 'stoploss-type' in the config.json:
    * 'standard' -> standard stoploss calculation based on the price when opening a trade (requires 'stoploss' percentage)
    * 'trailing' -> trailing stoploss function (requires 'stoploss' percentage)
    * 'dynamic' -> configurable stoploss function that can be changed in my_strategy_advanced.py

# Actions to be performed before this can be merged

- [ ] PR depends on updating the wiki. A lot of extra features are added in this PR, so the wiki needs to be up-to-date to match these changes.

# Security Measures

If 'dynamic' is enabled by the user but it is not defined in the users strategy, the system automatically chooses the standard stoploss configuration.

# Potential Issues/Future considerations

N/A

# Additional Info

N/A

### CODE REVIEW

**Comment by noudcorten** [comment]
# Fixes

Fixes #147 


# What has been changed?

* Restructured stoploss calculations
* Added trailing stoploss and dynamic stoploss
* Added 'stoploss-type' to config.json so users can switch between different types
* Added my_strategy_advanced.py to show how dynamic stoploss should be used


# Examples

* Users now have the choice to choose three different options as 'stoploss-type' in the config.json:
    * 'standard' -> standard stoploss calculation based on the price when opening a trade (requires 'stoploss' percentage)
    * 'trailing' -> trailing stoploss function (requires 'stoploss' percentage)
    * 'dynamic' -> configurable stoploss function that can be changed in my_strategy_advanced.py


# Actions to be performed before this can be merged

- [ ] PR depends on updating the wiki. A lot of extra features are added in this PR, so the wiki needs to be up-to-date to match these changes.


# Security Measures

N/A


# Potential Issues/Future considerations

The close price of the trade by the 'dynamic' stoploss is not 100% correctly calculated. As an example, currently in my_strategy_advanced.py the dynamic stoploss is defined as:
```python
dataframe.loc[
                (
                    (dataframe['low'] < dataframe['ema5'])
                ),
                'stoploss'] = 1
```
So when the low of the current candle is below the ema5, the stoploss is triggered. But, it could be the case that the low is far below the ema5. In live trading, when the low is at the ema5 line it instantly triggers the stoploss signal. But since we only work with candle data, we don't have this option. This was a problem with the 'standard' stoploss as well, but that could easily be fixed (in this PR) by using the given stoploss percentage to calculate back the value at which it should have been close. Here, we can not do this since the function is different per algo creater and it is impossible to calculate back at which value this trigger exactly happens, because it heavily depends on which indicators are used.

TL;DR dynamic stoploss closing price is not calculated 100% correct since it depends on the user's used indicators in his/her costum strategy.


# Additional Info

N/A

File: backtesting/strategy.py [hunk 1]
```diff
@@ -21,57 +21,54 @@ class Strategy(abc.ABC):
...
+        :param dataframe: dataframe filled with indicators from generate_indicators
+        :type dataframe: Dataframe
+        :return: dataframe filled with dynamic stoploss signals
+        :rtype: DataFrame
         """
         return
 
     @staticmethod
-    def change_timeframe(candle_data: DataFrame, new_timeframe: str) -> DataFrame:
+    def change_timeframe(dataframe: DataFrame, new_timeframe: str) -> DataFrame:
```
Comment (keebrev) [inline backtesting/strategy.py:71]: Remove this.

**Review by keebrev — APPROVED** [review]
